Source/ACE/Entity/WorldObject.cs

```diff
@@ -606,21 +638,91 @@ protected WorldObject(ObjectGuid guid, AceObject aceObject)
             PaletteGuid = aceObject.PaletteId;
         }
 
-        public void SetCombatMode(CombatMode newCombatMode)
+        public void SetCombatMode(CombatMode newCombatMode, Player player)
         {
-            log.InfoFormat("Changing combat mode for {0} to {1}", this.Guid, newCombatMode);
-            // TODO: any sort of validation
+            log.InfoFormat("Changing combat mode for {0} to {1}", Guid, newCombatMode);
+            // TODO: finish up this work.
+            if ((CombatMode == CombatMode.Missle) && (newCombatMode == CombatMode.Peace))
+            {
+                // delete the arrows from the world view
+            }
+            if ((CombatMode == CombatMode.Peace) && (newCombatMode == CombatMode.Missle))
+            {
+                // show the arrows to the world view
+            }
+
             CombatMode = newCombatMode;
             switch (CombatMode)
             {
                 case CombatMode.Peace:
                     SetMotionState(new UniversalMotion(MotionStance.Standing));
                     break;
                 case CombatMode.Melee:
-                    var gm = new UniversalMotion(MotionStance.UANoShieldAttack);
-                    gm.MovementData.CurrentStyle = (ushort)MotionStance.UANoShieldAttack;
-                    SetMotionState(gm);
+                    UniversalMotion gm = null;
+                    WorldObject meleeWeapon = null;
+
+                    // Let's see what if anything is equipped and if we have a shield.
+                    EquippedItem mEquipedMelee = player.Children.Find(s => s.EquipMask == EquipMask.MeleeWeapon);
+                    if (mEquipedMelee != null)
+                        meleeWeapon = player.GetInventoryItem(new ObjectGuid(mEquipedMelee.Guid));
```

Comment:
Are you sure you can access Player here (and other places within this function) safely (from a concurrency prospective)?